pkg/frontend/mysql_protocol.go

```diff
@@ -356,6 +362,315 @@ type response320 struct {
 	database      string
 }
 
+func (mp *MysqlProtocolImpl) SendPrepareResponse(stmt *PrepareStmt) error {
+	dcPrepare, ok := stmt.PreparePlan.GetDcl().Control.(*planPb.DataControl_Prepare)
+	if !ok {
+		return moerr.NewError(moerr.INTERNAL_ERROR, "can not get prepare plan in prepareStmt")
+	}
+	stmtID, err := GetPrepareStmtID(stmt.Name)
+	if err != nil {
+		return moerr.NewError(moerr.INTERNAL_ERROR, "can not get prepare stmtID")
+	}
+	paramTypes := dcPrepare.Prepare.ParamTypes
+	numParams := len(paramTypes)
+	columns := plan2.GetResultColumnsFromPlan(stmt.PreparePlan)
+	numColumns := len(columns)
+
+	// data := make([]byte, 12)
+	var data []byte
+	// status ok
+	data = append(data, 0)
+	// stmt id
+	data = mp.io.AppendUint32(data, uint32(stmtID))
+	// number columns
+	data = mp.io.AppendUint16(data, uint16(numColumns))
+	// number params
+	data = mp.io.AppendUint16(data, uint16(numParams))
+	// filter [00]
+	data = append(data, 0)
+	// warning count
+	data = append(data, 0, 0) // TODO support warning count
+	if err := mp.writePackets(data); err != nil {
+		return err
+	}
+
+	cmd := int(COM_STMT_PREPARE)
+	if numParams > 0 {
+		for i := 0; i < numParams; i++ {
+			column := new(MysqlColumn)
+			column.SetName("?")
+
+			err = convertEngineTypeToMysqlType(types.T(paramTypes[i]), column)
+			if err != nil {
+				return err
+			}
+			mp.SendColumnDefinitionPacket(column, cmd)
+		}
+
+		if err := mp.SendEOFPacketIf(0, 0); err != nil {
+			return err
+		}
+	}
+
+	if numColumns > 0 {
+		for i := 0; i < numColumns; i++ {
+			column := new(MysqlColumn)
+			column.SetName(columns[i].Name)
+
+			err = convertEngineTypeToMysqlType(types.T(columns[i].Typ.Id), column)
+			if err != nil {
+				return err
+			}
+			mp.SendColumnDefinitionPacket(column, cmd)
+		}
+
+		if err := mp.SendEOFPacketIf(0, 0); err != nil {
+			return err
+		}
+	}
+
+	return mp.flushOutBuffer()
+}
+
+func (mp *MysqlProtocolImpl) ParseExecuteData(stmt *PrepareStmt, data []byte, pos int) (names []string, vars []any, err error) {
+	dcPrepare, ok := stmt.PreparePlan.GetDcl().Control.(*planPb.DataControl_Prepare)
+	if !ok {
+		err = moerr.NewError(moerr.INTERNAL_ERROR, "can not get prepare plan in prepareStmt")
+		return
+	}
+	numParams := len(dcPrepare.Prepare.ParamTypes)
+
+	flag := data[pos]
+	pos++
+	if flag != 0 {
+		// TODO only support CURSOR_TYPE_NO_CURSOR flag now
+		err = moerr.NewError(moerr.INVALID_INPUT, fmt.Sprintf("unsupported flag %d", flag))
+		return
+	}
+
+	// skip iteration-count, always 1
+	pos += 4
+
+	if numParams > 0 {
+		var (
+			nullBitmaps []byte
+			paramValues []byte
+		)
+		nullBitmapLen := (numParams + 7) >> 3
+		if len(data) < (pos + nullBitmapLen + 1) {
+			err = moerr.NewError(moerr.INVALID_INPUT, "malform packet")
+			return
+		}
+		nullBitmaps = data[pos : pos+nullBitmapLen]
```

Comment:
I recommend that using readCountOfBytes here. It will check the buffer like the condition in line 460 has enough data or not.